### PR TITLE
Wip prefix search

### PR DESCRIPTION
**Pull request by kmlebedev**
fix quick list by prefix

### CODE REVIEW

**Comment by chrislusf** [comment]
If the underlying store does not natively support prefix search, how about just return an error UNSUPPORTED? And then, the filer store wrapper can implement the logic to filter by prefix.

This way seems more cleaner than reimplementing the same logic in those stores?

**Comment by kmlebedev** [comment]
ok

**Comment by kmlebedev** [comment]
@chrislusf  Do more clean

File: weed/filer2/filerstore.go [hunk 1]
```diff
@@ -129,6 +131,59 @@ func (fsw *FilerStoreWrapper) ListDirectoryEntries(ctx context.Context, dirPath
...
+	} else if err != nil {
+		return nil, err
+	}
+	for _, entry := range entries {
+		filer_pb.AfterEntryDeserialization(entry.Chunks)
+	}
+	return entries, nil
+}
+
+func (fsw *FilerStoreWrapper) ListDirectoryUnSupPrefixedEntries(ctx context.Context, dirPath util.FullPath, startFileName string, includeStartFile bool, limit int, prefix string) (entries []*Entry, err error) {
```
Comment (chrislusf) [inline weed/filer2/filerstore.go]: Not needed?

File: weed/filer2/cassandra/cassandra_store.go [hunk 2]
```diff
@@ -126,6 +125,10 @@ func (store *CassandraStore) DeleteFolderChildren(ctx context.Context, fullpath
 	return nil
 }
 
+func (store *CassandraStore) ListDirectoryPrefixedEntries(ctx context.Context, fullpath util.FullPath, startFileName string, inclusive bool, limit int, prefix string) (entries []*filer2.Entry, err error) {
+	return nil, fmt.Errorf("UNSUPPORTED")
```
Comment (chrislusf) [inline weed/filer2/cassandra/cassandra_store.go]: create a constant for the unsupported error

Comment (kmlebedev) [reply weed/filer2/cassandra/cassandra_store.go]: already

Comment (chrislusf) [reply weed/filer2/cassandra/cassandra_store.go]:
```
const UnsupportedError = fmt.Errorf("UNSUPPORTED")
```

Comment (kmlebedev) [reply weed/filer2/cassandra/cassandra_store.go]: > ```
> const UnsupportedError = fmt.Errorf("UNSUPPORTED")
> ```

I try
```
weed/filer2/filer.go:23:2: const initializer fmt.Errorf("UNSUPPORTED") is not a constant
```

Comment (chrislusf) [reply weed/filer2/cassandra/cassandra_store.go]: using var is fine

Comment (kmlebedev) [reply weed/filer2/cassandra/cassandra_store.go]: ready

Comment (chrislusf) [reply weed/filer2/cassandra/cassandra_store.go]: see https://golang.org/pkg/io/#pkg-variables

File: weed/filer2/filerstore.go [hunk 3]
```diff
@@ -138,7 +138,7 @@ func (fsw *FilerStoreWrapper) ListDirectoryPrefixedEntries(ctx context.Context,
 		stats.FilerStoreHistogram.WithLabelValues(fsw.ActualStore.GetName(), "list").Observe(time.Since(start).Seconds())
 	}()
 	entries, err := fsw.ActualStore.ListDirectoryPrefixedEntries(ctx, dirPath, startFileName, includeStartFile, limit, prefix)
-	if err.Error() == UnsupportedListDirectoryPrefixedErr {
+	if err.Error() == UnsupportedListDirectoryPrefixedErr.Error() {
```
Comment (chrislusf) [inline weed/filer2/filerstore.go]: no need to convert to string now.

File: weed/filer2/filerstore.go [hunk 4]
```diff
@@ -129,6 +131,54 @@ func (fsw *FilerStoreWrapper) ListDirectoryEntries(ctx context.Context, dirPath
...
+		}
+		if prefix == "" {
+			entries = notPrefixed
+		} else {
+			var lastFileName string
+			for count < limit {
+				for _, entry := range notPrefixed {
+					lastFileName = entry.Name()
+					if strings.HasPrefix(entry.Name(), prefix) {
+						count++
```
Comment (chrislusf) [inline weed/filer2/filerstore.go:155]: the count could go over the limit here already.

File: weed/filer2/filerstore.go [hunk 5]
```diff
@@ -129,6 +131,54 @@ func (fsw *FilerStoreWrapper) ListDirectoryEntries(ctx context.Context, dirPath
...
+						entries = append(entries, entry)
+					}
+				}
+				if count >= limit {
+					break
+				}
+
+				notPrefixed, err = fsw.ActualStore.ListDirectoryEntries(ctx, dirPath, lastFileName, includeStartFile, limit)
```
Comment (chrislusf) [inline weed/filer2/filerstore.go]: in this line, includeStartFile should be false?

Comment (chrislusf) [inline weed/filer2/filerstore.go]: the limit can be `limit - count`

with this, we don't need to check whether the count is over limit in line 155.

File: weed/filer2/filerstore.go (hunk 4 above)
Comment (kmlebedev) [reply weed/filer2/filerstore.go:155]: ok

File: weed/filer2/filerstore.go (hunk 5 above)
Comment (kmlebedev) [reply weed/filer2/filerstore.go]: ok

Comment (kmlebedev) [reply weed/filer2/filerstore.go]: Its seemes limit is pagination limit and it makes no sense to reduce by count

**Comment by chrislusf** [comment]
Merged! Thanks!

**Pull request merged**